### PR TITLE
Feature/sg 953 generalise dataset logic

### PR DESCRIPTION
**Pull request by Louis-Dupont**
Now that we are starting to have more datasets (coco detection and coco/voc/BDDD segmentation are coming), I had a lot of duplicate code, and I think that it makes sense to build a general abstraction `BaseImageLabelDirectoryDataset ` that actually covers the majority of the datasets (open for suggestions on a better name)

### CODE REVIEW

File: src/data_gradients/datasets/FolderProcessor.py [hunk 1]
```diff
@@ -103,66 +112,15 @@ def _match_file_names(self, all_images_file_names: List[str], all_labels_file_na
...
-            label_extensions=label_extensions,
-            image_extensions=image_extensions,
-            verbose=verbose,
-        )
-
-    def get_image_and_label_file_names(self, images_dir: str, labels_dir: str) -> List[Tuple[str, str]]:
-        """Gather all image and label files that are both listed in the config_path and in the directories.
-        :param images_dir:      The directory containing the images.
-        :param labels_dir:      The directory containing the labels.
-        :return:                A list of tuple(<path-to-image>, <path-to-label>).
```
Comment (Louis-Dupont) [inline src/data_gradients/datasets/FolderProcessor.py:162]: I realized that in lots of cases I wrote the following code:
```python
        if config_path is None:
            self.image_label_tuples = ImageLabelFilesIterator(
                images_dir=os.path.join(root_dir, images_subdir),
                labels_dir=os.path.join(root_dir, labels_subdir),
                image_extensions=image_extensions,
                label_extensions=label_extensions,
                verbose=verbose,
            )
        else:
            self.image_label_tuples = ImageLabelConfigIterator(
                images_dir=os.path.join(root_dir, images_subdir),
                labels_dir=os.path.join(root_dir, labels_subdir),
                config_path=os.path.join(root_dir, config_path),
                image_extensions=image_extensions,
                label_extensions=label_extensions,
                verbose=verbose,
            )
```
This made me realize that we should not have two classes for this, but just one.
The downside is that this class is starting to be relatively big which is never good, but I think it is reasonable

File: src/data_gradients/datasets/base_dataset.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    def __getitem__(self, index: int) -> Tuple[np.ndarray, np.ndarray]:
+        ...
+
+
+class BaseImageLabelDirectoryDataset(BaseDataset, ABC):
+    """Base class for any dataset that is primarily made of an image and label directories."""
+
+    def __init__(
+        self,
+        *,
```
Comment (BloodAxe) [inline src/data_gradients/datasets/base_dataset.py:40]: What `*` is here for?

File: src/data_gradients/datasets/base_dataset.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+from abc import ABC, abstractmethod
+from typing import Tuple, Sequence, Optional
+
+import numpy as np
+from torch.utils.data.dataset import Dataset
+from data_gradients.datasets.FolderProcessor import ImageLabelFilesIterator, DEFAULT_IMG_EXTENSIONS
+from data_gradients.datasets.utils import load_image, ImageChannelFormat
+
+
+class BaseDataset(Dataset, ABC):
```
Comment (BloodAxe) [inline src/data_gradients/datasets/base_dataset.py:11]: Let's imagine we are adding NLP task at some point. Would that base class be suitable for this task with `load_image` and `load_labels` methods that returns numpy arrays?

Second complain I have is that API is not clear at all.
* Why there is `__iter__` and `__getitem__`? Which one should I use (prefer)?
* Why there are separate methods `load_image` and `load_labels` and why one have to implement them explicitly if he just implemented `__getitem__`?

Third problem is that we don't use ImageSample / SegmentationSample / DetectionSample and instead use "plain" numpy arrays without any metadata and clear structure what is stored inside such arrays.

File: src/data_gradients/datasets/detection/voc_format_detection_dataset.py [hunk 4]
```diff
@@ -154,25 +153,20 @@ def __init__(
...
-    def load_image(self, index: int) -> np.ndarray:
-        img_file, _ = self.image_label_tuples[index]
-        return load_image(path=img_file, channel_format=ImageChannelFormat.RGB)
-
-    def load_labels(self, index: int) -> np.ndarray:
-        _, label_path = self.image_label_tuples[index]
+    def load_labels(self, path: str) -> np.ndarray:
 
-        with open(label_path) as f:
+        with open(path) as f:
```
Comment (BloodAxe) [inline src/data_gradients/datasets/detection/voc_format_detection_dataset.py:169]: I suggest to add `encoding="utf-8"` or wrap with try/catch to ensure we don't crash if for some XML would contain unicode symbols.

File: src/data_gradients/datasets/detection/yolo_format_detection_dataset.py [hunk 5]
```diff
@@ -135,25 +129,11 @@ def load_labels(self, index: int) -> np.ndarray:
                 try:
                     labels.append(list(map(float, lines_elements)))
                 except ValueError as e:
-                    raise ValueError(
-                        f"Invalid label: {line} from {label_path}.\nExpected 5 elements (class_id, cx, cy, w, h), got {len(lines_elements)}."
-                    ) from e
+                    raise ValueError(f"Invalid label: {line} from {path}.\nExpected 5 elements (class_id, cx, cy, w, h), got {len(lines_elements)}.") from e
```
Comment (BloodAxe) [inline src/data_gradients/datasets/detection/yolo_format_detection_dataset.py:132]: I think this is brittle design. Some YOLO annotation files also contains keypoint annotations in the same txt file next to bounding boxes. This code would crash on such files. 

Can we extract standalone "read_yolo_annotation(lines: List[str], on_parse_error='raise')" method that we can convert with unit tests to ensure it:

* Can handle spaces, tabs and other whitespace symbols without crashing
* Would skip empty line or commented lines (Starting with `#`). Yes, it is a thing in some YOLO annotations as well
* Check that indeed the first element is a label (it's an integer and not float)
* If `on_parse_error='raise'` then it would raise is unexpected string comes in and if `on_parse_error='warining'` it would emit a warning and skip this line.

**Review by BloodAxe — CHANGES_REQUESTED** [review]
See the inline comments. IMO this PR has some stuff to improve

File: src/data_gradients/datasets/detection/yolo_format_detection_dataset.py (hunk 5 above)
Comment (Louis-Dupont) [reply src/data_gradients/datasets/detection/yolo_format_detection_dataset.py]: What do you think about this ?
98af9d2d44a5a4616c2e095793f35565a9436117

File: src/data_gradients/datasets/detection/voc_format_detection_dataset.py (hunk 4 above)
Comment (Louis-Dupont) [reply src/data_gradients/datasets/detection/voc_format_detection_dataset.py]: [ddbc2a8](https://github.com/Deci-AI/data-gradients/pull/131/commits/ddbc2a8ced958f21f988333b3f873d459b8ffe84)

File: src/data_gradients/datasets/base_dataset.py (hunk 3 above)
Comment (Louis-Dupont) [reply src/data_gradients/datasets/base_dataset.py:11]: Well, yeah I guess this abstract class is not that useful anyway, I think we can just take it out.

About `load_image` and `load_labels` there is a difference, which is that `load_image` can be easily be implemented for one class and be shared among any child (which is what I did), which means that we only need to implement `load_labels` afterward. Besides, having it separated to `__getitem__` was motivated to avoid mixing responsabilities;  for instance if we want to add a transform step, this should not be part of `load_labels`, but should be done inside `__getitems__` instead. But I agree that this should not be part of the most basic abstract class.

Concerning the samples, whether we should return these objects or not is a good question.
My approach to this was "Let's implement simple datasets similar to what users may use (return image/label) so that users can also play around with it easily", also motivated by the simplicity and low coupling this would create with the rest of the code. But I also see clear advantages to what you proposed:

Pro:
- More explicit output
- (In theory) no need to go through the `batch_processor` steps. In practice, we would need to update the design for that, because currently any dataset output is passed to the `batch_processor`.
- Abstracting the output of the dataset into these `Sample` classes makes these datasets easy to use in different context without needing to worry how the dataset outputs data.

Con:
- The current pipeline expects a dataset/dataloader/generator that yields samples in dict/list/tuple/... format, not dataclass. To resolve this, I think the clean way would be to check if the user dataset is a DG dataset; If yes, take the dataset output as is, if not we wrap the user dataset with the `batch_processor`. Introducing a clean "wrapping" class  would mean more work because the current pipelines are a bit coupled with the `batch_processor`.
- [Most importantly] Not compatible with dataloaders (dict/tuple/list can be collated, not dataclasses). We can still define custom collate functions, but it is definitely less nice for users

Any other thing that you think should be considered? And don't you think that the dataloader compatibility is a bit problematic for a `Dataset` ?

File: src/data_gradients/datasets/detection/yolo_format_detection_dataset.py (hunk 5 above)
Comment (BloodAxe) [reply src/data_gradients/datasets/detection/yolo_format_detection_dataset.py]: Good!

File: src/data_gradients/datasets/base_dataset.py (hunk 3 above)
Comment (BloodAxe) [reply src/data_gradients/datasets/base_dataset.py:11]: Ok, I now have even more questions that before..

> About load_image and load_labels there is a difference, which is that load_image can be easily be implemented for one class and be shared among any child (which is what I did), which means that we only need to implement load_labels afterward. 

Composition is better than inheritance.

> (In theory) no need to go through the batch_processor steps. In practice, we would need to update the design for that, because currently any dataset output is passed to the batch_processor.

Like, yes, updating design is a right thing if design is broken


> And don't you think that the dataloader compatibility is a bit problematic for a Dataset ?

But why we have to use dataloader at all in the first place here? This is somewhat making task unnecessary more complicated to me. If we already have the dataset class - just iterate over it and that's it. 

For who this new class is for? Is it for use, for internal API or we're going to ask user to inherit from this class? If feel that this is a part of internal API, but at the same time I'm not sure about it.

**Review by BloodAxe — DISMISSED** [review]
LGTM
